File: api/jobspy.py
import json
import math
from datetime import date, datetime
from http.server import BaseHTTPRequestHandler
from urllib.parse import parse_qs, urlparse

from jobspy import scrape_jobs


def value(row, name):
    raw = row.get(name, "")
    if raw is None or (isinstance(raw, float) and math.isnan(raw)):
        return ""
    if isinstance(raw, (datetime, date)):
        return raw.isoformat()
    return str(raw).strip()


def fetch(source, term, location, hours_old):
    if source == "google":
        listings = scrape_jobs(
            site_name=["google"],
            google_search_term=f"{term} jobs in {location} posted in the last day",
            results_wanted=25,
            verbose=0,
        )
    else:
        listings = scrape_jobs(
            site_name=[source],
            search_term=term,
            location=location,
            country_indeed="India",
            hours_old=hours_old,
            results_wanted=25,
            verbose=0,
        )

    jobs = []
    for _, row in listings.fillna("").iterrows():
        url = value(row, "job_url")
        title = value(row, "title")
        if not url or not title:
            continue
        jobs.append({
            "url": url,
            "title": title,
            "company": value(row, "company") or "Unknown",
            "company_logo": value(row, "company_logo"),
            "location": value(row, "location"),
            "description": value(row, "description"),
            "job_type": value(row, "job_type"),
            "date_posted": value(row, "date_posted"),
            "min_amount": value(row, "min_amount"),
            "max_amount": value(row, "max_amount"),
            "currency": value(row, "currency"),
            "interval": value(row, "interval"),
        })
    return jobs


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        query = parse_qs(urlparse(self.path).query)
        source = query.get("source", ["indeed"])[0]
        if source not in {"indeed", "glassdoor", "google"}:
            self._send(400, {"error": "Unsupported JobSpy source."})
            return
        try:
            jobs = fetch(
                source,
                query.get("keywords", ["software engineer"])[0],
                query.get("location", ["India"])[0],
                max(1, int(query.get("hoursOld", ["24"])[0])),
            )
            self._send(200, jobs)
        except Exception as error:
            self._send(502, {"error": str(error)})

    def do_POST(self):
        self.do_GET()

    def _send(self, status, payload):
        body = json.dumps(payload).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

File: lib/jobspy.ts
import { createHash } from "node:crypto";
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { resolve } from "node:path";
import type { Job } from "./types";

const run = promisify(execFile);
const sources = new Set(["indeed", "glassdoor", "google"]);
type Listing = { url: string; title: string; company: string; company_logo: string; location: string; description: string; job_type: string; date_posted: string; min_amount: string; max_amount: string; currency: string; interval: string };

const iso = (value: string) => value && !Number.isNaN(Date.parse(value)) ? new Date(value).toISOString() : null;
const salary = (job: Listing) => [job.min_amount, job.max_amount].filter(Boolean).join(" - ") + (job.currency ? ` ${job.currency}` : "") + (job.interval ? ` / ${job.interval}` : "") || null;

export async function fetchJobSpyJobs(source: string, term = "software engineer", location = "India", lookbackSeconds = 86_400, workerUrl?: string): Promise<Job[]> {
  if (!sources.has(source)) throw new Error("Unsupported JobSpy source.");
  const hoursOld = Math.max(1, Math.round(lookbackSeconds / 3_600));
  let listings: Listing[];
  if (process.env.VERCEL) {
    const endpoint = new URL(workerUrl || `https://${process.env.VERCEL_URL}/api/jobspy`);
    endpoint.search = new URLSearchParams({ source, keywords: term, location, hoursOld: String(hoursOld) }).toString();
    const response = await fetch(endpoint, { cache: "no-store", signal: AbortSignal.timeout(90_000) });
    const body = await response.json() as Listing[] | { error?: string };
    if (!response.ok) throw new Error((body as { error?: string }).error || `${source} worker returned ${response.status}.`);
    listings = body as Listing[];
  } else {
    const { stdout } = await run(process.env.PYTHON_BIN || "python", [resolve(process.cwd(), "scripts", "jobspy_worker.py"), source, term, location, String(hoursOld)], { timeout: 90_000, maxBuffer: 2_000_000 });
    listings = JSON.parse(stdout) as Listing[];
  }
  if (!Array.isArray(listings) || !listings.length) throw new Error(`${source} returned no jobs.`);
  const fetchedAt = new Date().toISOString();
  return listings.map((job) => {
    const postedAt = iso(job.date_posted);
    return { id: `${source}:${createHash("sha256").update(job.url).digest("hex")}`, source, externalId: job.url, canonicalUrl: job.url, contentHash: createHash("sha256").update([job.title, job.company, job.location, job.description, postedAt].join("\u0000")).digest("hex"), title: job.title, company: job.company, companyLogoUrl: job.company_logo || null, location: job.location || null, country: /india/i.test(job.location) ? "India" : "India", category: null, jobType: job.job_type || null, salary: salary(job), description: job.description || null, postedAt, fetchedAt, active: true } satisfies Job;
  });
}
